### PR TITLE
fix name filter of the list images API

### LINKED ISSUE #18181: Filtered list images API doesn't work with registry port
**Description of problem:**
The [list images API](http://docs.docker.com/engine/reference/api/docker_remote_api_v1.21/#list-images) seems to not work when adding the `filter` parameter that includes a registry with a port.

**`docker version`:**
```
Client:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.5.1
 Git commit:   76d6bc9
 Built:        Tue Nov  3 21:18:00 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   76d6bc9
 Built:        Tue Nov  3 19:20:09 UTC 2015
 OS/Arch:      linux/amd64
```

**`docker info`:**
```
Containers: 3
Images: 275
Server Version: 1.9.0
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 291
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.12-boot2docker
Operating System: Boot2Docker 1.9.0 (TCL 6.4); master : 16e4a2a - Tue Nov  3 19:49:22 UTC 2015
CPUs: 1
Total Memory: 996.1 MiB
Name: dockci
ID: DAXI:SNXH:NRG3:2ZHQ:KO6R:6BPW:KNLU:QZ74:TRU7:46DS:YRPU:NAMK
Debug mode (server): true
 File Descriptors: 32
 Goroutines: 66
 System Time: 2015-11-23T12:36:51.105258529Z
 EventsListeners: 0
 Init SHA1:
 Init Path: /usr/local/bin/docker
 Docker Root Dir: /mnt/sda1/var/lib/docker
Username: thatpanda
Registry: https://index.docker.io/v1/
Labels:
 provider=vmwarefusion
```

**`uname -a`:**
```
Darwin Rickys-MacBook-Pro.local 15.0.0 Darwin Kernel Version 15.0.0: Wed Oct 28 11:41:13 PDT 2015; root:xnu-3248.20.42~5/RELEASE_X86_64 x86_64
```
```
Linux dockci 4.1.12-boot2docker #1 SMP Tue Nov 3 06:03:36 UTC 2015 x86_64 GNU/Linux
```

**Environment details (AWS, VirtualBox, physical, etc.):**
Docker Machine with VMWare Fusion provider running on OSX El Capitan

**How reproducible:**
100%, across docker CLI, docker-py, and raw HTTP

**Steps to Reproduce:**
- Dockerfile containing:
```
FROM scratch
ADD Dockerfile /
```

- `docker build -t just-built .`
- `docker tag just-built:latest localhost:5000/just-built:latest`
- `docker images`
```
REPOSITORY                          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
localhost:5000/just-built           latest              a5e832e1966b        13 hours ago        30 B
just-built                          latest              a5e832e1966b        13 hours ago        30 B
[ ... removed more images ... ]
```

- `docker images just-built`
```
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
just-built          latest              a5e832e1966b        13 hours ago        30 B
```

- `docker images 'localhost:5000/just-built'`
```
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
```

_From here on is just more examples_

**Using raw HTTP**

- `http --verify no --cert $DOCKER_CERT_PATH/cert.pem --cert-key $DOCKER_CERT_PATH/key.pem "https://$(docker-machine ip $(docker-machine active)):2376/images/json?filter=just-built"` (where the `http` command is HTTPie)
```
HTTP/1.1 200 OK
Content-Length: 202
Content-Type: application/json
Date: Tue, 24 Nov 2015 01:46:54 GMT
Server: Docker/1.9.0 (linux)

[
    {
        "Created": 1448281661,
        "Id": "a5e832e1966bad9ca5821738ff70f37d737ec79779c98fe36ef477ad603d8c9d",
        "Labels": null,
        "ParentId": "",
        "RepoDigests": [],
        "RepoTags": [
            "just-built:latest"
        ],
        "Size": 30,
        "VirtualSize": 30
    }
]
```
- _no escape_ `http --verify no --cert $DOCKER_CERT_PATH/cert.pem --cert-key $DOCKER_CERT_PATH/key.pem "https://$(docker-machine ip $(docker-machine active)):2376/images/json?filter=localhost:5000/just-built"`
- _: escaped_ `http --verify no --cert $DOCKER_CERT_PATH/cert.pem --cert-key $DOCKER_CERT_PATH/key.pem "https://$(docker-machine ip $(docker-machine active)):2376/images/json?filter=localhost%3A5000/just-built"`
- _: and / escaped_ `http --verify no --cert $DOCKER_CERT_PATH/cert.pem --cert-key $DOCKER_CERT_PATH/key.pem "https://$(docker-machine ip $(docker-machine active)):2376/images/json?filter=localhost%3A5000%2Fjust-built"`
```
HTTP/1.1 200 OK
Content-Length: 3
Content-Type: application/json
Date: Tue, 24 Nov 2015 01:48:08 GMT
Server: Docker/1.9.0 (linux)

[]
```

**Using docker-py**

- `python -c 'import docker; k = docker.utils.kwargs_from_env(); k["tls"].assert_hostname = False; c = docker.Client(**k); import pprint; pprint.pprint(c.images(name="just-built"))'`
```
[{'Created': 1448281661,
  'Id': 'a5e832e1966bad9ca5821738ff70f37d737ec79779c98fe36ef477ad603d8c9d',
  'Labels': None,
  'ParentId': '',
  'RepoDigests': [],
  'RepoTags': ['just-built:latest'],
  'Size': 30,
  'VirtualSize': 30}]
```

- ` python -c 'import docker; k = docker.utils.kwargs_from_env(); k["tls"].assert_hostname = False; c = docker.Client(**k); import pprint; pprint.pprint(c.images(name="localhost:5000/just-built"))'`
```
[]
```

**Actual Results:**
No results returned for `localhost:5000/just-built`, where results returned for `just-built`

**Expected Results:**
Exactly the same results returned for `localhost:5000/just-built`, as results returned for `just-built`

### PR DESCRIPTION
**Pull request by mountkin**
Fixes #18181

### CODE REVIEW

**Comment by runcom** [comment]
just tested it was working in docker 1.8.3 and it works with this patch

LGTM

ping @aaronlehmann @tonistiigi

**Comment by vdemeester** [comment]
LGTM 🐯

**Comment by mountkin** [comment]
There's another potential bug with the image filter.

```bash
docker tag ubuntu hello:v1
docker tag ubuntu hello:v1v1
docker images hello:v1
```
screenshot from 2015-11-24 18 09 20

I just wrote another patch to fix the problem. Should I add another commit to this PR or open another PR when this one is merged?

**Comment by vdemeester** [comment]
@mountkin I'm not sure about that one but it doesn't feel consistent. 
Looking at the actual `docker image` commands, to list all `ubuntu` image we do `docker image "ubuntu*"`, so I would guess `docker image "hello:v1*"` do the same.

I guess it could be in that one too (but that would mean both @runcom and I would have to re-LGTM it :stuck_out_tongue_closed_eyes:).

**Comment by mountkin** [comment]
@vdemeester Perhaps I didn't describe it clearly. I meant that the `docker images hello:v1` command should not show the image `hello:v1v1`.
To make the review simpler I think it's better to open a new PR when this one is merged :smile_cat:

**Comment by runcom** [comment]
I think another commit here is fine (these seems small fixes) I'll test it as soon as you push it. 

Do you happen to discover what caused all of this though?

File: graph/list.go [hunk 1]
```diff
@@ -88,11 +82,12 @@ func (s *TagStore) Images(filterArgs, filter string, all bool) ([]*types.Image,
 				}
 			}
 		}
-		for ref, id := range repository {
-			imgRef := utils.ImageReference(repoName, ref)
-			if !strings.Contains(imgRef, filterTagName) {
```
Comment (mountkin) [inline graph/list.go]: @runcom This `strings.Contains` led to the problem

Comment (runcom) [inline graph/list.go]: I mean which commit broke this

Comment (mountkin) [inline graph/list.go]: It was added in https://github.com/docker/docker/pull/13501

Comment (runcom) [inline graph/list.go]: oh ok found https://github.com/docker/docker/commit/4fb88d2e11f2f16c520b36e4c6e00db44ae26b8c mmm @vdemeester may tell more then

Comment (vdemeester) [inline graph/list.go]: oh, it's mine :sweat_smile:

**Comment by vdemeester** [comment]
> @vdemeester Perhaps I didn't describe it clearly. I meant that the docker images hello:v1 command should not show the image hello:v1v1.

@mountkin Yeah I do agree. I was just pointing at that we should also add support for `*`, `docker images hello:v1*` should show hello:v1 and hello:v1v1.

**Comment by mountkin** [comment]
@vdemeester I like the idea but it's out of the scope of this PR :goat:

**Comment by mountkin** [comment]
@vdemeester @runcom  another commit added, PTAL, thanks

**Comment by vdemeester** [comment]
@mountkin sounds fair on the `*` stuff.

re LGTM 🐯 then :wink:

**Comment by calavera** [comment]
@mountkin please, squash your commits. There is no need of a second commit when it's a full unit of work.

Please, do not merge before #17924 lands on master.

**Comment by mountkin** [comment]
@calavera in fact this PR is fixing 2 bugs (#18181 and https://github.com/docker/docker/pull/18187#issuecomment-159225462), so I think fixing them in separate commit is much clearer.
Now that  #17924 is merged, it seems the 2 problems are all gone. I'll close this.
I migrated the test case in #18215, please help to review it. Thanks

**Pull request closed without merging**